### DISCUSSION TITLE
Error: Cookies can only be modified in a Server Action or Route Handler.

### DISCUSSION THREAD

**Discussion by rmartin93** [vercel/next.js#68325] [Help] [answered]
### Summary

I have a file with "use server" at the top, and I'm somehow getting an error saying I can only modify cookies in server actions. Is there some factor other than being in a server action that I have to keep in mind? 

I have a comment in the code below highlighting which line is causing the error. The getUser function is another server action that goes and gets the user's info from our database. 

```ts
"use server";
export async function getUserSession(): Promise<{
    user: User | null;
    error?: string;
}> {
    try {
        const userCookie = await getCookie("userData");
        if (!userCookie || userCookie.value === "") {
            const user = await getUser();
            // This is the line where the error happens 
            setCookie("userData", JSON.stringify(user));
        }
        const userData: User =
            userCookie && userCookie.value !== ""
                ? JSON.parse(userCookie.value)
                : null;
        return { user: userData };
    } catch (error) {
        console.error(
            "Error fetching user session on app > _actions > cookies.ts > getUserSession()!",
            error,
        );
        return {
            user: null,
            error: "Error fetching user session on app > _actions > cookies.ts > getUserSession()!",
        };
    }
}
```

```ts
// Set cookie function in the same file 
export async function setCookie(name: string, value: string) {
    cookies().set(name, value, {
        httpOnly: true,
        secure: process.env.APP_ENV === "Production",
        maxAge: 60 * 60 * 24 * 7, // 1 week
        path: "/",
    });
}
```

**Comment by icyJoseph** [comment]
How are you calling this function.

**Reply by rmartin93** [reply]
I was calling it in my app/page.tsx file. 

```ts
import { getUserSession } from "./_actions/cookies";
export default async function Home() {

    const { user } = await getUserSession();
    if (!user) return;

    return (
        <div>
            <div className="mb-5">Welcome back</div>
        </div>
    );
}
```

It's interesting that you ask how I'm calling this function. 

As long as the setCookie is in a "use server" file, I thought I should be fine. I call server actions from server and client components all the time and I've never had issues unless I forgot to put "use server" at the top. 

There seems to be some piece of information about server actions or (more likely) cookies that I don't understand.

**Reply by rmartin93** [reply]
I thought it might having something to do with this line from the documentation, which I do not understand if I'm being honest.

`HTTP does not allow setting cookies after streaming starts` 

But if you can think of something else I'm totally open to it. 

I just want to know why this doesn't work, when all my other server actions seem to work perfectly.

**Reply by icyJoseph** [reply, marked as answer]
Right, a server action, formally, is only a server action, when invoked from the client, not during SSR.

You are calling it as you render in the server, and that is somewhat equivalent to, inlining the function, that is, calling the set cookie while streaming has started. 

`use server` is a way to tell the bundler, that the client will have the ability, through some mechanism, to trigger a function server side, through a network call. So, while it is possible to call a function, marked to be a server action, during render, it doesn't make it a server action call, just a function call.

**Reply by rmartin93** [reply]
Okay, so this is causing an error because I'm trying to set a cookie during SSR (which I guess would be during "streaming"), right? 

Again, just trying to make sure I understand. 

Thank you so much for your help!

**Reply by icyJoseph** [reply]
Yes, in general, setting the cookies while rendering a server component has this kind of issue.

**Reply by rmartin93** [reply]
Awesome, thank you so much!

**Comment by zwelhtetyan** [comment]
My case is that user manually change the token in cookie and request the resource from the external api endpoint. Server return 401 status and I want to catch in ky interceptor and delete token from cookie. I also try calling route handler, and I noticed this didn't even call. I really confuse how that work. Can anyone help me to remove cookie in the right way?

```tsx
export const api = ky.create({
  prefixUrl: BASE_URL,
  hooks: {
    beforeRequest: [injectTokenBeforeRequest],
    beforeError: [formatError],
  },
})

// =========================================================================== //

export const formatError: BeforeErrorHook = async (error) => {
  const { response } = error

  const isUnauthorized = response.status === 401

  if (isUnauthorized) {
  // This line goes the error "Cookies can only be modified in a Server Action or Route Handler"
    await deleteSession()
  }

  const isJsonContent = response.headers
    .get("content-type")
    ?.includes("application/json")

  if (isJsonContent) return response.json()

  return error
}
```

```tsx
"use server"

export const deleteSession = async () => {
  try {
    const cookieStore = await cookies()
    cookieStore.delete(COOKIE_NAME)
  } catch (err) {
    throw new Error("Failed to delete session")
  }
}

```
